src/ray/core_worker/actor_manager.h

```diff
@@ -14,34 +14,122 @@
 
 #pragma once
 
+#include "absl/container/flat_hash_map.h"
 #include "ray/core_worker/actor_handle.h"
+#include "ray/core_worker/reference_count.h"
+#include "ray/core_worker/transport/direct_actor_transport.h"
 #include "ray/gcs/redis_gcs_client.h"
 
 namespace ray {
 
-// Interface for testing.
-class ActorManagerInterface {
- public:
-  virtual void PublishTerminatedActor(const TaskSpecification &actor_creation_task) = 0;
-
-  virtual ~ActorManagerInterface() {}
-};
-
 /// Class to manage lifetimes of actors that we create (actor children).
 /// Currently this class is only used to publish actor DEAD event
 /// for actor creation task failures. All other cases are managed
 /// by raylet.
-class ActorManager : public ActorManagerInterface {
+class ActorManager {
  public:
-  ActorManager(gcs::ActorInfoAccessor &actor_accessor)
-      : actor_accessor_(actor_accessor) {}
+  explicit ActorManager(
+      std::shared_ptr<gcs::GcsClient> gcs_client,
+      std::shared_ptr<CoreWorkerDirectActorTaskSubmitterInterface> direct_actor_submitter,
+      std::shared_ptr<ReferenceCounterInterface> reference_counter)
+      : gcs_client_(gcs_client),
+        direct_actor_submitter_(direct_actor_submitter),
+        reference_counter_(reference_counter) {}
+
+  ~ActorManager() {}
+
+  friend class ActorManagerTest;
+
+  /// Register an actor handle.
+  ///
+  /// This should be called when an actor handle is given to us by another task
+  /// or actor. This may be called even if we already have a handle to the same
+  /// actor.
+  ///
+  /// \param[in] actor_handle The actor handle.
+  /// \param[in] outer_object_id The object ID that contained the serialized
+  /// actor handle, if any.
+  /// \param[in] caller_id The caller's task ID
+  /// \param[in] call_site The caller's site.
+  /// \return The ActorID of the deserialized handle.
+  ActorID RegisterActorHandle(std::unique_ptr<ActorHandle> actor_handle,
+                              const ObjectID &outer_object_id, const TaskID &caller_id,
+                              const std::string &call_site,
+                              const rpc::Address &caller_address);
+
+  /// Get a handle to an actor.
+  ///
+  /// \param[in] actor_id The actor handle to get.
+  /// \return reference to the actor_handle's pointer.
+  /// NOTE: Returned actorHandle should not be stored anywhere.
+  std::unique_ptr<ActorHandle> &GetActorHandle(const ActorID &actor_id);
+
+  /// Check if an actor handle that corresponds to an actor_id exists.
+  /// \param[in] actor_id The actor id of a handle.
+  /// \return True if the actor_handle for an actor_id exists. False otherwise.
+  bool CheckActorHandleExists(const ActorID &actor_id);
 
-  /// Called when an actor that we own can no longer be restarted.
-  void PublishTerminatedActor(const TaskSpecification &actor_creation_task) override;
+  /// Give this worker a handle to an actor.
```

Comment:
I think it's more accurate to say that this creates an actor handle. I would actually rename this method to `CreateActorHandle` and make `AddActorHandle` a private method.

Reply:
Oh hmm I see that this is also called by the registration for named actor handles. Can we have a separate method for each of the types of handles?
1. Initial creation
2. Passing actor handle as an arg
3. Named actor lookup

Reply:
`Passing actor handle as an arg` is already implemented as `RegisterActorHandle`. I think I can make this private and create methods for 1 and 3. 

Btw, `CreatingActorHandle` might be misleading because actor handle is created before this function is called. What about

```
Initial creation => AddNewActorHandle
Passing actor handle as an arg => RegisterActorHandle(/*is_detached*/false)
Named actor lookup => RegisterActorHandle(/*is_detached*/true)

private method
AddActorHandle => This will still have is_owner_handle because it is used for `SetDeleteCallback`, but I will add TODO to delete this when I remove the legacy actor management path.
```

Reply:
Please check out the interface and lmk if you satisfy with it!